addons/binding/org.openhab.binding.bluetooth.ruuvitag/src/main/java/org/openhab/binding/bluetooth/ruuvitag/handler/RuuviTagHandler.java

```diff
@@ -0,0 +1,163 @@
+                fieldPresent |= updateStateIfLinked(RuuviTagBindingConstants.CHANNEL_ID_HUMIDITY,
+                        ruuvitagData.getHumidity(), SmartHomeUnits.PERCENT);
+                fieldPresent |= updateStateIfLinked(RuuviTagBindingConstants.CHANNEL_ID_MEASUREMENT_SEQUENCE_NUMBER,
+                        ruuvitagData.getMeasurementSequenceNumber(), SmartHomeUnits.ONE);
+                fieldPresent |= updateStateIfLinked(RuuviTagBindingConstants.CHANNEL_ID_MOVEMENT_COUNTER,
+                        ruuvitagData.getMovementCounter(), SmartHomeUnits.ONE);
+                fieldPresent |= updateStateIfLinked(RuuviTagBindingConstants.CHANNEL_ID_PRESSURE,
+                        ruuvitagData.getPressure(), SIUnits.PASCAL);
+                fieldPresent |= updateStateIfLinked(RuuviTagBindingConstants.CHANNEL_ID_TEMPERATURE,
+                        ruuvitagData.getTemperature(), SIUnits.CELSIUS);
+                fieldPresent |= updateStateIfLinked(RuuviTagBindingConstants.CHANNEL_ID_TX_POWER,
+                        ruuvitagData.getTxPower(), DECIBEL_MILLIWATTS);
+                if (fieldPresent) {
+                    updateStatus(ThingStatus.ONLINE);
+                } else {
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                            "Received Ruuvi Tag data but no fields could be parsed");
+                }
+            } else {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                        "Received bluetooth data which could not be parsed to any known Ruuvi Tag data formats");
+            }
+        } else {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                    "Received bluetooth data with empty ");
+        }
+        super.onScanRecordReceived(scanNotification);
+    }
+
+    /**
+     * Update channel state
+     *
+     * Update is not done when value is null.
+     *
+     * @param channel channel id
+     * @param value   value to update
+     * @param unit    unit associated with value, or null when we are updating DecimalType
+     * @return whether the value was present
+     */
+    private <T extends Quantity<T>> boolean updateStateIfLinked(String channel, @Nullable Number value,
+            @Nullable Unit<T> unit) {
+        if (value == null) {
+            return false;
+        }
+        ChannelUID channelUID = getChannelUID(channel);
+        if (!isLinked(channelUID)) {
+            return true;
+        }
+        if (unit == null) {
+            updateState(channelUID, new DecimalType(value.doubleValue()));
+        } else {
+            updateState(channelUID, new QuantityType<>(value, unit));
+        }
+        return true;
+    }
+
+    private ChannelUID getChannelUID(String channelID) {
+        return channelCache.computeIfAbsent(channelID, id -> new ChannelUID(getThing().getUID(), id));
```

Comment:
Instead of keeping a cache I would suggest to change the setting of data to iterate over the channels itself. With `getThing().getChannels()` the list of channels is available. So instead of creating a channel uid and cache it, just iterate over the channels and in a switch statement handle the match between channel. Something line this:
```
import static org.openhab.binding.bluetooth.ruuvitag.RuuviTagBindingConstants.*;

...
            for (Channel channel : getThing().getChannels()) {
                switch (channel.getUID().getId()) {
                    case CHANNEL_ID_ACCELERATIONX:
                        fieldPresent |= updateStateIfLinked(channel, ruuvitagData.getAccelerationX(), STANDARD_GRAVITY); 
                        break;
                    case .
                        
                }
            }
```

Reply:
The caching is to avoid constructing `ChannelUID` which is surprisingly expensive operation on low-end machines such as raspberry pi. 

Your suggestion sounds like a sensible approach -- will fix it

Reply:
Fixed in f4da827